### PR TITLE
added unsqueeze_dim to apply_rotary_pos_emb

### LINKED ISSUE #26948: Making the unsqueeze dimension parameterized in the apply_rotary_pos_emb function in modeling_llama.py
### Feature request

Hi Huggingface Team, 
We would like to request to make the 'unsqueeze(1)' in [these two lines ](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L209C1-L210C41) parameterized. To be precise, we would like to request that the following lines

```
def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
    cos = cos[position_ids].unsqueeze(1) 
    sin = sin[position_ids].unsqueeze(1)
    ...
```
be converted to something like the following
```
def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
    cos = cos[position_ids].unsqueeze(unsqueeze_dim)
    sin = sin[position_ids].unsqueeze(unsqueeze_dim)
    ...
```

### Motivation

We are trying to import and use the [apply_rotary_pos_emb function](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L208) and the [LlamaRotaryEmbedding class](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L119). However, our query and key tensors have the shape `[batch_size, sequence_len, heads, dim]`. To make them compatible with the `apply_rotary_pos_emb` function, we have to transpose the tensors to `[batch_size, heads, sequence_len, dim]`, call `apply_rotary_pos_emb` on them and then transpose the tensors back to `[batch_size, sequence_len, heads, dim]`. These unnecessary transposes could be avoided if the `apply_rotary_pos_emb` function had a parameter which controlled the dimension on which the `unsqueeze`'s [here](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L209C1-L210C41) were applied. 

_Please note that the Llama Huggingface code also [does similar back and forth transposes](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L442C1-L463C52), and hence could benefit from this **very small** code change as well._

### Your contribution

We are willing to submit a PR for this.

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
# Making the unsqueeze dimension parameterized in the apply_rotary_pos_emb function in modeling_llama.py 

This PR introduces a new parameter to the [apply_rotary_pos_emb function](https://github.com/huggingface/transformers/blob/08a2edfc6629a323effd7a85feafed9e6701e2dd/src/transformers/models/llama/modeling_llama.py#L208) which allows the specification of the dimension along which to unsqueeze to make the cosine and sine rotary tensors broadcastable to the query and key tensors. This will make the function compatible with codebases that have different shapes for the query and key tensors without needing any back-and-forth transposing.

Fixes #26948

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [Link to the issue](https://github.com/huggingface/transformers/issues/26948)
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@gante , @ArthurZucker

### CODE REVIEW

**Review by gante — APPROVED** [review]
Thank you for opening the PR @ShashankMosaicML! And nice touch, the docstring :)

CI is failing because of the `# Copied from ...` statement, which is used by our internal tools and CI to maintain one file per model policy while enabling the propagation of bugfixes and upgrades. 

In this particular case, there is also a bug on our end -- recently we've made Llama the base implementation, from which others are copied from. As such, you'll have to:
1. Remove the `# Copied from ...` statement in L194
2. Run make `fix-copies` in your terminal
3. Push the changes

You may also need to run `make fixup` afterwards, which applies automatic code formatting.

**Comment by ShashankMosaicML** [comment]
Thank you @gante for reviewing and providing me the steps to fix the errors! However I still see some errors after following the steps, and on clicking the details for the error, I see the following. Could you please let me know how to fix this error? Thank you!
Screenshot 2023-10-28 at 10 16 16 PM

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

Change to the function LGTM - there's just some addition changes in this diff to docstrings which need to be removed before merging.

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -191,10 +191,21 @@ def rotate_half(x):
     return torch.cat((-x2, x1), dim=-1)
 
 
-# Copied from transformers.models.gpt_neox.modeling_gpt_neox.apply_rotary_pos_emb
-def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
-    cos = cos[position_ids].unsqueeze(1)  # [seq_len, dim] -> [batch_size, 1, seq_len, head_dim]
-    sin = sin[position_ids].unsqueeze(1)
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply src/transformers/models/llama/modeling_llama.py:195]: :)

File: src/transformers/models/beit/image_processing_beit.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ class BeitImageProcessor(BaseImageProcessor):
         size (`Dict[str, int]` *optional*, defaults to `{"height": 256, "width": 256}`):
             Size of the output image after resizing. Can be overridden by the `size` parameter in the `preprocess`
             method.
-        resample (`PILImageResampling`, *optional*, defaults to `Resampling.BICUBIC`):
+        resample (`PILImageResampling`, *optional*, defaults to `3`):
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply src/transformers/models/beit/image_processing_beit.py]: Got it, I have done these changes and pushed the changes.

File: src/transformers/models/blip/image_processing_blip.py [hunk 3]
```diff
@@ -53,7 +53,7 @@ class BlipImageProcessor(BaseImageProcessor):
         size (`dict`, *optional*, defaults to `{"height": 384, "width": 384}`):
             Size of the output image after resizing. Can be overridden by the `size` parameter in the `preprocess`
             method.
-        resample (`PILImageResampling`, *optional*, defaults to `Resampling.BICUBIC`):
+        resample (`PILImageResampling`, *optional*, defaults to `3`):
```
[the remark this answers is not in the document]
Comment (ShashankMosaicML) [reply src/transformers/models/blip/image_processing_blip.py]: Got it, I implemented these changes and have pushed them.

**Comment by ShashankMosaicML** [comment]
Hi @amyeroberts , thanks for suggesting the changes. I have incorporated those, but  some quality checks are still failing. Could you take a look?
Screenshot 2023-10-30 at 2 10 24 PM

**Comment by gante** [comment]
@ShashankMosaicML running `make fixup` then committing the changes doesn't fix it?

**Comment by ShashankMosaicML** [comment]
@gante , I think that worked! (I wasn't running `make fixup` properly earlier 😅 )

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating!
